src/main/java/org/rocstreaming/roctoolkit/ReferenceCollector.java

```diff
@@ -0,0 +1,99 @@
+package org.rocstreaming.roctoolkit;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Collect open {@link NativeObjectReference} for avoid being garbage collected.
+ *
+ * @param <T>       the type of {@link NativeObjectReference} to be collected.
+ */
+class ReferenceCollector<T extends NativeObjectReference> implements Iterable<T> {
+
+    /**
+     * References map indexed by their native pointers.
+     */
+    private final Map<Long, T> references;
+
+    /**
+     * Create a new <code>ReferenceCollector</code>.
+     */
+    ReferenceCollector() {
+        references = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Add a new reference to the collection.
+     *
+     * @param reference         reference to add.
+     */
+    void add(final T reference) {
+        references.put(reference.getPtr(), reference);
+    }
+
+    /**
+     * Remove a reference from the collection.
+     *
+     * @param reference         reference to remove.
+     */
+    void remove(final T reference) {
+        references.remove(reference.getPtr());
+    }
+
+    /**
+     * Get the number of references inside the collection.
+     *
+     * @return      number of references.
+     */
+    int size() {
+        return references.size();
+    }
+
+    /**
+     * Iterator over <code>ReferenceCollector</code> iterating references with dependencies
+     * ({@link NativeObjectReference#dependsOn}) firstly and secondly the references without
+     * dependencies.
+     */
+    class ReferenceCollectorIterator implements Iterator<T> {
+
+        /**
+         * Iterator over references with dependencies.
+         */
+        private Iterator<T> withDependencies;
+
+        /**
+         * Iterator over references without dependencies.
+         */
+        private Iterator<T> withoutDependencies;
+
+        /**
+         * Create a new <code>ReferenceCollectorIterator</code>.
+         */
+        ReferenceCollectorIterator() {
+            withDependencies = references.values().parallelStream()
+                            .filter(r -> r.getDependsOn().isPresent())
+                            .iterator();
+            withoutDependencies = references.values().parallelStream()
+                            .filter(r -> !r.getDependsOn().isPresent())
+                            .iterator();
+        }
+
+        @Override
+        public boolean hasNext() {
+            return withDependencies.hasNext() || withoutDependencies.hasNext();
+        }
+
+        @Override
+        public T next() {
+            if (withDependencies.hasNext()) return withDependencies.next();
+            if (withoutDependencies.hasNext()) return withoutDependencies.next();
+            return null;
+        }
```

Comment:
We currently have only two layer of dependencies (receiver/sender on first layer, context on second), but I see that ReferenceCollector is designed as if it supported arbitrary number of layers.

However, will it work correctly if there are more than two layers?

If `A` depends on `B`, and `B` depends on `C`, then `A` and `B` will be both in `withDependencies` list. Is it guaranteed that `A` is always before `B` in this list?

If yes, then could you please add brief comment?

If no, then I'd suggest of the following:

* make the implementation correct even on multi-layer dependencies;
* OR: make the limitations of the implementation somehow explicit from the code; e.g. maybe hold to maps (with and without dependencies) and require that the dependency itself can be only in without-dependencies map.

Or, if we decide to remove close-on-shutdown feature, the iterator can be probably removed at all.

Reply:
With this implementation the iterator doesn't order the `withDependencies` objects, so in case of multiple levels of depencies like your example it's not guaranteed that `A` is always before `B`.
I can do a more general solution simply by storing the references in a directed graph and then the iterator ordering will be a result of a topological sort on the graph.

Reply:
added to sender/receiver link to context. So context wouldn't be collected by CG before sender/reciever. With such approach, we don't need a directed graph. Fixed in https://github.com/roc-streaming/roc-java/pull/40